### PR TITLE
[CFM-1274] add wiki select

### PR DESCRIPTION
**Pull request by anvmn**
https://issuetracker.amplexor.com/jira/browse/CFM-1274

### CODE REVIEW

File: project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module [hunk 1]
```diff
@@ -95,3 +95,72 @@ function c4m_og_group_menu_page_access() {
...
+    '#options' => $wiki_details,
+    '#weight' => -1,
+  );
+
+  // If path provided, set selected Wiki Page as a default value.
+  if (in_array($form['link_path']['#default_value'], array_keys($wiki_details))) {
+    $form['wiki_select']['#default_value'] = $form['link_path']['#default_value'];
+  }
+
+  $form['link_title']['#weight'] = -3;
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module]: Add comments

File: project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module [hunk 2]
```diff
@@ -95,3 +95,72 @@ function c4m_og_group_menu_page_access() {
 
   return og_menu_access('node', $gid);
 }
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module]: Add docs

File: project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module [hunk 3]
```diff
@@ -95,3 +95,72 @@ function c4m_og_group_menu_page_access() {
...
+  if ($form['link_path']['#type'] == 'value') {
+    return;
+  }
+
+  $tree = menu_build_tree($group->book['menu_name']);
+  // Resolve pages under the node of the menu tree.
+  $top_pages = array_shift($tree)['below'];
+
+  // Default select option.
+  $wiki_details = array('' => t('--- SELECT ---'));
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module]: => `select a wiki`

Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module]: `$wiki_details` => `$wiki_options`

File: project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module [hunk 4]
```diff
@@ -95,3 +95,72 @@ function c4m_og_group_menu_page_access() {
 
   return og_menu_access('node', $gid);
 }
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function c4m_og_group_menu_form_og_menu_edit_item_form_alter(&$form, &$form_state, $form_id) {
+  $gid = og_menu_get_context();
+  $group = node_load($gid);
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module]: use `c4m_og_current_group()` instead

File: project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module [hunk 5]
```diff
@@ -95,3 +95,72 @@ function c4m_og_group_menu_page_access() {
...
+    return;
+  }
+
+  // The select element of the wiki pages should not be displayed on a default
+  // menu link.
+  if ($form['link_path']['#type'] == 'value') {
+    return;
+  }
+
+  $tree = menu_build_tree($group->book['menu_name']);
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module]: Is it possible that a project won't have a book ?

File: project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module [hunk 6]
```diff
@@ -95,3 +95,72 @@ function c4m_og_group_menu_page_access() {
...
+  // Resolve pages under the node of the menu tree.
+  $top_pages = array_shift($tree)['below'];
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module]: what are `$top_pages` stand for ?
the comment is not clear.

File: project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module [hunk 7]
```diff
@@ -95,3 +95,72 @@ function c4m_og_group_menu_page_access() {
...
+  $form['link_title']['#weight'] = -3;
+  $form['link_path']['#weight'] = -2;
+
+  drupal_add_js(drupal_get_path('module', 'c4m_og_group_menu') . '/js/wiki-select.js');
+}
+
+/**
+ * Recursive function.
+ *
+ * Generates 'options' input, for 'Wiki Page' select element.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module]: This should be the first line.

File: project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module [hunk 8]
```diff
@@ -95,3 +95,72 @@ function c4m_og_group_menu_page_access() {
...
+  }
+
+  $form['link_title']['#weight'] = -3;
+  $form['link_path']['#weight'] = -2;
+
+  drupal_add_js(drupal_get_path('module', 'c4m_og_group_menu') . '/js/wiki-select.js');
+}
+
+/**
+ * Recursive function.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module]: This should be part of the explanation of the function, please move it down and elaborate the logic.

File: project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module [hunk 9]
```diff
@@ -95,3 +95,72 @@ function c4m_og_group_menu_page_access() {
...
+ * Generates 'options' input, for 'Wiki Page' select element.
+ *
+ * @param array $page
+ *   Details of current page, and pages bellow it.
+ * @param array $wiki_details
+ *   Contains details (lint => title) of all wiki pages.
+ */
+function _c4m_og_group_menu_process_page($page, &$wiki_details) {
+  // Add details of current page to wiki details.
+  $page_link  = $page['link'];
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module]: remove extra space before the `=`

File: project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/js/wiki-select.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+/**
+ * @file
+ * React on wiki page select.
+ */
+
+(function ($) {
+
+  Drupal.behaviors.wikiSelect = {
+    attach: function (context) {
+      $("#edit-wiki-select").on("change", function() {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/js/wiki-select.js]: Add docs.

File: project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module (hunk 5 above)
Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module]: Don't think so, but will check this anyway...

File: project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module [hunk 11]
```diff
@@ -95,3 +95,73 @@ function c4m_og_group_menu_page_access() {
...
+  // If path is provided, set selected Wiki Page as a default value.
+  if (in_array($form['link_path']['#default_value'], array_keys($wiki_options))) {
+    $form['wiki_select']['#default_value'] = $form['link_path']['#default_value'];
+  }
+
+  // Place wiki title and wiki path fields above select element.
+  $form['link_title']['#weight'] = -3;
+  $form['link_path']['#weight'] = -2;
+
+  drupal_add_js(drupal_get_path('module', 'c4m_og_group_menu') . '/js/wiki-select.js');
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module]: Let's rename the file to `c4m-wiki-select.js`

File: project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module [hunk 12]
```diff
@@ -95,3 +95,73 @@ function c4m_og_group_menu_page_access() {
...
+}
+
+/**
+ * Generates 'options' input, for 'Wiki Page' select element.
+ *
+ * @param array $page
+ *   Details of current page, and pages bellow it.
+ * @param array $wiki_options
+ *   Contains options (lint => title) of all wiki pages.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og_group_menu/c4m_og_group_menu.module]: what `lint` means in here ?

**Comment by anvmn** [comment]
out

**Pull request merged**